pkg/asset/installconfig/gcp/validation.go

```diff
@@ -336,3 +337,86 @@ func validateCredentialMode(client API, ic *types.InstallConfig) field.ErrorList
+		} else {
+			errMsg := checkArchitecture(image.Architecture, ic.ControlPlane.Architecture, "controlPlane")
+			if errMsg != "" {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("platform", "gcp", "defaultMachinePlatform", "osImage"), osImage, errMsg))
+			}
+			for _, compute := range ic.Compute {
+				errMsg = checkArchitecture(image.Architecture, compute.Architecture, "compute")
+				if errMsg != "" {
+					allErrs = append(allErrs, field.Invalid(field.NewPath("platform", "gcp", "defaultMachinePlatform", "osImage"), osImage, errMsg))
+				}
+			}
+		}
+	}
+
+	if ic.ControlPlane != nil && ic.ControlPlane.Platform.GCP != nil && ic.ControlPlane.Platform.GCP.OSImage.Name != "" {
+		osImage := ic.ControlPlane.Platform.GCP.OSImage
+		image, err := client.GetImage(context.TODO(), osImage.Name, osImage.Project)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("controlPlane", "platform", "gcp", "osImage"), osImage, fmt.Sprintf(errorMessage, err)))
+		} else {
+			errMsg := checkArchitecture(image.Architecture, ic.ControlPlane.Architecture, "controlPlane")
+			if errMsg != "" {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("controlPlane", "platform", "gcp", "osImage"), osImage, errMsg))
+			}
+		}
+	}
+
+	for idx, compute := range ic.Compute {
+		if compute.Platform.GCP == nil {
+			continue
+		}
+		fieldPath := field.NewPath("compute").Index(idx)
+		osImage := compute.Platform.GCP.OSImage
+		if osImage.Name != "" {
+			image, err := client.GetImage(context.TODO(), osImage.Name, osImage.Project)
+			if err != nil {
+				allErrs = append(allErrs, field.Invalid(fieldPath.Child("platform", "gcp", "osImage"), osImage, fmt.Sprintf(errorMessage, err)))
+			} else {
+				errMsg := checkArchitecture(image.Architecture, compute.Architecture, "compute")
+				if errMsg != "" {
+					allErrs = append(allErrs, field.Invalid(fieldPath.Child("platform", "gcp", "osImage"), osImage, errMsg))
+				}
+			}
+		}
+	}
+
+	return allErrs
+}
+
+func checkArchitecture(imageArch string, icArch types.Architecture, location string) string {
+	const unspecifiedArch string = "ARCHITECTURE_UNSPECIFIED"
+	// The possible architecture names from image.Architecture are of type string hence we cannot directly obtain the possible values
+	// In the docs the possible values are ARM64, X86_64, and ARCHITECTURE_UNSPECIFIED
+	// There is no simple translation between the architecture values from Google and the architecture names used in the install config so a map is used
+	var (
+		translateArchName = map[string]types.Architecture{
+			"ARM64":         types.ArchitectureARM64,
+			"X86_64":        types.ArchitectureAMD64,
+			unspecifiedArch: "",
+		}
+	)
+
+	if translateArchName[imageArch] != icArch {
+		if imageArch == unspecifiedArch {
+			logrus.Warn(fmt.Sprintf("Boot image architecture is unspecified and might not be compatible with %s %s nodes", icArch, location))
```

Comment:
If you do this check first, you don't need to add `unspecifiedArch` to the map. You can even immediately return.